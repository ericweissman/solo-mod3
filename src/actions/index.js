export const isLoading = (bool) => ({
  type: 'IS_LOADING',
  isLoading: bool
})

export const hasErrored = (error) => ({
  type: 'HAS_ERRORED',
  error
})

export const fetchArtistsSuccess = (artists, genre) => ({
  type: 'FETCH_ARTISTS_SUCCESS',
  artists,
  genre
})

export const getDeepHouseSuccess = (artists) => ({
  type: 'GET_DEEP_HOUSE_SUCCESS',
  artists
})

export const getFutureBassSuccess = (artists) => ({
  type: 'GET_FUTURE_BASS_SUCCESS',
  artists
})

export const getGlitchHopSuccess = (artists) => ({
  type: 'GET_GLITCH_HOP_SUCCESS',
  artists
})

export const getTrapSuccess = (artists) => ({
  type: 'GET_TRAP_SUCCESS',
  artists
})

export const addToFavorites = (artist) => ({
  type: "ADD_TO_FAVORITES",
  artist
})

export const removeFromFavorites = (id) => ({
  type: "REMOVE_FROM_FAVORITES",
  id
})

export const populateFavorites = (artists) => ({
  type: 'POPULATE_FAVORITES',
  artists
})

