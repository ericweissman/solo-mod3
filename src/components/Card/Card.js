import React from 'react'
import YouTube from 'react-youtube';

const Card = ({name, video, wiki}) => {
  return(
    <div>
      <h4>{name}</h4>
      <p>{video}</p>
      <YouTube 
        videoId={video}
        title={name}
      />
      <a href=`${wiki}`>
    </div>
  )
}

export default Card