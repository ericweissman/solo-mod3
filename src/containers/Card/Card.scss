.card {
  display: flex;
  flex-direction: column;
  align-items: center;
  border: 2px solid $color-main;
  border-radius: 5px;
  margin: 15px;
  width: 300px;
  height: 335px;;
  background-color: $color-main;
  box-shadow: 10px 10px 5px -7px rgba(0,0,0,0.75);
  &:hover {
    animation: grow 0.5s ease 0s 1 normal;
    animation-fill-mode: forwards;
  }
  .video {
    border: 2px solid $color-background-main;    
  }
  h4 {
    font-family: $primary-font;
    color: $text-white;
    font-size: 25px;
    padding: 15px 0;
  }
  i {
    padding-right: 5px;
  }
  a {
    font-family: $secondary-font;
    color: $color-accent;
    background-color: $color-background-main;
    height: 30px;
    width: 120px;
    text-align: center;
    line-height: 30px;
    text-decoration: none;
    margin: 15px 0;
    border-radius: 20px;
    &:hover {
      background-color: $color-accent;
      color: $color-main;
    }
  }
  button {
    @extend a;
    margin: 0 0 15px 0;
    border: none;
    outline: none;
  }
.favorites {
  width: 140px;
  margin: 0 0 15px 0;
  }
}

@keyframes grow {
  0% {
    transform: scale(1)
  }
  100% {
    transform: scale(1.01)
  }
}

@media only screen and (max-width: 560px) {
  .card {
    width: 285px;
  }
}