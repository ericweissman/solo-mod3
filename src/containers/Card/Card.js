import React, { Component } from 'react'
import YouTube from 'react-youtube';
import { connect } from 'react-redux'
import PropTypes from 'prop-types'
import { addToFavorites, removeFromFavorites } from '../../actions'

export class Card extends Component {
  handleFavorite = (e) => {
    const { artist, favorites } = this.props
    const { favorited } = this.props.artist
    e.preventDefault();
    let updatedFavorites
    if (!favorited) {
      
      this.props.addToFavorites(artist)
      updatedFavorites = [...favorites, artist]
    } else if (favorited) {
      this.props.removeFromFavorites(artist)
      updatedFavorites = favorites.filter(favorite => favorite.id !== artist.id)
    }
    localStorage.setItem('favorites', JSON.stringify(updatedFavorites))
  }

  render() {
    const { name, id, wiki, favorited } = this.props.artist
    const opts = {
      height: '152',
      width: '270'
    }

    return (
      <div className='card'>
        <h4>{name}</h4>
        <YouTube
          className='video'
          videoId={id}
          title={name}
          opts={opts}
        />
        <a href={wiki} target='blank'> <i class="fas fa-info-circle"></i>More Info</a>
        {
          favorited ?
            <button className='favorites' onClick={this.handleFavorite}><i class="fas fa-times-circle"></i>Remove from Favorites</button> :
            <button className='favorites' onClick={this.handleFavorite}><i class="fas fa-plus-circle"></i>Add to Favorites</button>
        }
      </div>
    )
  }
}

export const mapStateToProps = (state) => ({
  favorites: state.favorites,
})

export const mapDispatchToProps = (dispatch) => ({
  addToFavorites: (artist) => dispatch(addToFavorites(artist)),
  removeFromFavorites: (artist) => dispatch(removeFromFavorites(artist))
})

export default connect(mapStateToProps, mapDispatchToProps)(Card)

Card.propTypes = {
  favorites: PropTypes.array,
  addToFavorites: PropTypes.func,
  removeFromFavorites: PropTypes.func,
}


Card.defaultProps = {
  favorites: [],
}