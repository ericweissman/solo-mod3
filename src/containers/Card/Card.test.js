import React from 'react'
import { shallow } from 'enzyme'
import { Card, mapDispatchToProps, mapStateToProps } from '../Card/Card'
import { addToFavorites, removeFromFavorites } from '../../actions'

describe('Card', () => {
  let wrapper
  const mockArtist = {
    name: 'PL',
    id: 'abc123',
    wiki: 'somelink',
    favorited: false
  }

  const mockArtistFavorited = {
    name: 'PL',
    id: 'abc123',
    wiki: 'somelink',
    favorited: true
  }

  const mockProps = {
    addToFavorites: jest.fn(),
    removeFromFavorites: jest.fn(),
    favorites: []
  }

  it('should match the correct snapshot', () => {
    wrapper = shallow(<Card artist={mockArtist} {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  it('should match the correct snapshot if the artist has been favorited', () => {
    wrapper = shallow(<Card artist={mockArtistFavorited} {...mockProps} />)
    expect(wrapper).toMatchSnapshot()
  })

  describe('handleFavorite', () => {
    it('should add artist to favorites if the artist has not been favorited', () => {
      wrapper = shallow(<Card artist={mockArtist} {...mockProps} />)
      let favorites = ['asasf']
      localStorage.setItem('favorites', JSON.stringify(favorites))
      const mockEvent = { preventDefault: jest.fn() }
      wrapper.instance().handleFavorite(mockEvent)
      expect(mockProps.addToFavorites).toHaveBeenCalledWith(mockArtist)
      localStorage.setItem('favorites', JSON.stringify(favorites.concat(mockArtist.id)))
      const itemsInStorage = JSON.parse(localStorage.getItem('favorites')).length
      expect(itemsInStorage).toEqual(2)
    })


    it('should remove the artist from favorites if the artist has not been favorited before', () => {
      wrapper = shallow(<Card artist={mockArtistFavorited} {...mockProps} />)
      const mockEvent = { preventDefault: jest.fn() }
      wrapper.instance().handleFavorite(mockEvent)
      expect(mockProps.removeFromFavorites).toHaveBeenCalledWith(mockArtistFavorited)
    })
  })


  describe('mapStateToProps', () => {
    it('should return an object with all of the genres', () => {
      const mockState = {
        favorites: [],
        testMSTP: 'fake',
        testTwo: 'also fake'
      }
      const expected = {
        favorites: [],
      }

      const mappedState = mapStateToProps(mockState)
      expect(mappedState).toEqual(expected)
    })
  })

  describe('mapDispatchToProps', () => {
    it('should call dispatch with addToFavorites', () => {
      const mockDispatch = jest.fn()
      const mockArtist = { name: 'PL', id: '123abc', wiki: 'someurl', favorited: false }
      const actionToDispatch = addToFavorites(mockArtist)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.addToFavorites(mockArtist)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })

    it('should call dispatch with removeFromFavorites', () => {
      const mockDispatch = jest.fn()
      const mockArtist = { name: 'PL', id: '123abc', wiki: 'someurl', favorited: false }
      const actionToDispatch = removeFromFavorites(mockArtist)

      const mappedProps = mapDispatchToProps(mockDispatch)
      mappedProps.removeFromFavorites(mockArtist)
      expect(mockDispatch).toHaveBeenCalledWith(actionToDispatch)
    })
  })
})