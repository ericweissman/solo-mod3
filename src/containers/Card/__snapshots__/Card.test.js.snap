// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card should match the correct snapshot 1`] = `
<div>
  <h4>
    PL
  </h4>
  <YouTube
    className={null}
    containerClassName=""
    id={null}
    onEnd={[Function]}
    onError={[Function]}
    onPause={[Function]}
    onPlay={[Function]}
    onPlaybackQualityChange={[Function]}
    onPlaybackRateChange={[Function]}
    onReady={[Function]}
    onStateChange={[Function]}
    opts={Object {}}
    title="PL"
    videoId="abc123"
  />
  <a
    href="somelink"
    target="blank"
  >
    More Info
  </a>
  <button
    onClick={[Function]}
  >
    Add to Favorites
  </button>
</div>
`;

exports[`Card should match the correct snapshot if the artist has been favorited 1`] = `
<div>
  <h4>
    PL
  </h4>
  <YouTube
    className={null}
    containerClassName=""
    id={null}
    onEnd={[Function]}
    onError={[Function]}
    onPause={[Function]}
    onPlay={[Function]}
    onPlaybackQualityChange={[Function]}
    onPlaybackRateChange={[Function]}
    onReady={[Function]}
    onStateChange={[Function]}
    opts={Object {}}
    title="PL"
    videoId="abc123"
  />
  <a
    href="somelink"
    target="blank"
  >
    More Info
  </a>
  <button
    onClick={[Function]}
  >
    Delete from favorites
  </button>
</div>
`;
